### PR TITLE
Move the repo to newer version of FxCop and Roslyn analyzers (2.3.0-b…

### PR DESCRIPTION
**Pull request by mavasani**
…eta1)

**Customer scenario**

Dogfood newer version of Roslyn and FXCop analyzers (based on IOperation).

**Bugs this fixes:**

N/A

**Workarounds, if any**

N/A

**Risk**

Low, this is just an infrastructure change.

**Performance impact**

None

**Is this a regression from a previous update?**

N/A

**Root cause analysis:**

N/A

**How was the bug found?**

N/A

### CODE REVIEW

**Comment by mavasani** [comment]
Tagging @dotnet/analyzer-ioperation @dotnet/roslyn-infrastructure for review.

**Comment by genlu** [comment]
Can we re-enable the rules listed in #18689?

**Comment by sharwell** [comment]
@mavasani I just checked out the source branch and I'm not seeing any analyzers in the IDE. Does this require a build newer than 15.3 preview 1 to work?

**Comment by mavasani** [comment]
@sharwell I have the following VS and the CA diagnostics do show up in the IDE:

```
Microsoft Visual Studio Enterprise 2017 Int Preview (2)
Version 15.3.0-pre.1.0+26524.0.d15rel
VisualStudio.15.IntPreview/15.3.0-pre.1.0+26524.0.d15rel
Microsoft .NET Framework
Version 4.6.01586
```

However, I note a ton of new CA diagnostics reported in the error list, but they don't seem to be generated during build. I had assumed the new diagnostics I enabled were not firing for our code, but seems like these analyzers are not running on build. Investigating now..

**Comment by mavasani** [comment]
@sharwell - You are correct. The diagnostics I was seeing in the IDE were coming from VSIX installed due to my local build of analyzers repo. The PR in the current state doesn't include the analyzers as we changed the props file folder in the NuGet package (I will also add an MSBuild error which will be fired when the analyzers/props file are not found, so they are not silently skipped in build). When I enabled them locally, they fire over 500+ CA warnings :). I'll clean up the ones that I can right now, and open tracking issues for the ones which have too many active instances.

File: build/Rulesets/Roslyn_BuildRules.ruleset [hunk 1]
```diff
@@ -36,15 +36,15 @@
...
+    <Rule Id="CA1801" Action="Warning" />
+    <Rule Id="CA1802" Action="Warning" />
+    <Rule Id="CA1804" Action="Warning" />
     <Rule Id="CA1806" Action="None" /> <!-- "do not ignore results": we do this in many places, for example fire-and-forget async or HRESULTs we don't care about -->
     <Rule Id="CA1812" Action="None" /> <!-- disabled as we create many internal types via reflection-based mechanisms -->
     <Rule Id="CA1814" Action="None" /> <!-- prefer jagged arrays to multidimensional: a silly piece of advice -->
     <Rule Id="CA1815" Action="None" />
-    <Rule Id="CA1821" Action="None" /> <!-- broken due to https://github.com/dotnet/roslyn-analyzers/issues/1176 -->
-    <Rule Id="CA1823" Action="None" /> <!-- "avoid unused private fields": many issues including https://github.com/dotnet/roslyn-analyzers/issues/933 -->
+    <Rule Id="CA1821" Action="Warning" />
```
Comment (alrz) [inline build/Rulesets/Roslyn_BuildRules.ruleset]: :question: why there is no 1822? also, do these mean that the fixes also show up in the IDE?

**Comment by mavasani** [comment]
This is now blocked on https://github.com/dotnet/roslyn/issues/19819

File: build/Rulesets/Roslyn_BuildRules.ruleset (hunk 1 above)
Comment (mavasani) [inline build/Rulesets/Roslyn_BuildRules.ruleset]: >  why there is no 1822?

CA1822 is not enabled by default in the FXCop analyzers implementation. However, you are correct in the sense that we should have an explicit entry for it with Action="None" in this ruleset. I will add this entry.

> do these mean that the fixes also show up in the IDE

Yes, for all the rules that have fixers implemented in the analyzers package, the fixers will light up in the editor.

Comment (mavasani) [inline build/Rulesets/Roslyn_BuildRules.ruleset]: See https://github.com/dotnet/roslyn-analyzers/blob/ce77e9a81de80385aae9e635f9084bc22672e561/src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/MarkMembersAsStatic.cs#L32

**Comment by genlu** [comment]
@mavasani Can we enable the following analyzers as well (all 3 were warnings)?

    RS0012
    RS0014
    CA1825

**Comment by mavasani** [comment]
@genlu Yes, there is bunch of work still needed for this PR. I'll enable those as well.

**Comment by sharwell** [comment]
@mavasani Can you add me back as a reviewer when this is ready for review?

File: src/VisualStudio/Core/Def/Implementation/Workspace/GlobalUndoServiceFactory.WorkspaceGlobalUndoTransaction.cs [hunk 2]
```diff
@@ -117,11 +117,13 @@ public void Dispose()
                 }
             }
 
+#pragma warning disable CA1821 // Remove empty Finalizers
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/Workspace/GlobalUndoServiceFactory.WorkspaceGlobalUndoTransaction.cs]: Tracked with https://github.com/dotnet/roslyn-analyzers/issues/1241

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/InvisibleEditor.cs [hunk 3]
```diff
@@ -145,9 +145,11 @@ public void Dispose()
             }
         }
 
+#pragma warning disable CA1821 // Remove empty Finalizers
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/InvisibleEditor.cs]: Tracked with https://github.com/dotnet/roslyn-analyzers/issues/1241

File: build/Targets/Settings.props [hunk 4]
```diff
@@ -15,7 +15,7 @@
     <VSIXExpInstallerPackageName>RoslynTools.Microsoft.VSIXExpInstaller</VSIXExpInstallerPackageName>
     <VSIXExpInstallerPackageVersion>0.2.1-beta</VSIXExpInstallerPackageVersion>
     <RoslynDiagnosticsNugetPackageVersion>$(MicrosoftNetRoslynDiagnosticsVersion)</RoslynDiagnosticsNugetPackageVersion>
-    <RoslynDiagnosticsPropsFilePath>$(NuGetPackageRoot)\Microsoft.Net.RoslynDiagnostics\$(RoslynDiagnosticsNugetPackageVersion)\build\Microsoft.Net.RoslynDiagnostics.props</RoslynDiagnosticsPropsFilePath>
+    <RoslynDiagnosticsPropsFilePath>$(NuGetPackageRoot)\Microsoft.Net.RoslynDiagnostics\$(RoslynDiagnosticsNugetPackageVersion)\roslyn\Microsoft.Net.RoslynDiagnostics.props</RoslynDiagnosticsPropsFilePath>
```
Comment (genlu) [inline build/Targets/Settings.props]: Just curious, why the change in path? I thought `build` is the default location for props/targets files in NuGet package.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/InvisibleEditor.cs (hunk 3 above)
Comment (nguerrera) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/InvisibleEditor.cs]: See my comment there, finalizers with only asserts should have #if !DEBUG around them. Only Contract.Requires should have #if !CONTRACTS_FULL. The analysis is correctly identifying a perf issue when these calls get compiled away.

File: src/VisualStudio/Core/Def/Implementation/Workspace/GlobalUndoServiceFactory.WorkspaceGlobalUndoTransaction.cs (hunk 2 above)
Comment (nguerrera) [inline src/VisualStudio/Core/Def/Implementation/Workspace/GlobalUndoServiceFactory.WorkspaceGlobalUndoTransaction.cs]: Ditto.

File: build/Targets/Settings.props (hunk 4 above)
Comment (mavasani) [inline build/Targets/Settings.props]: See https://github.com/dotnet/roslyn-analyzers/pull/1197#issue-228560961.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/InvisibleEditor.cs (hunk 3 above)
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/InvisibleEditor.cs]: I added a comment here: https://github.com/dotnet/roslyn-analyzers/issues/1241#issuecomment-310740456. I'll fix the Roslyn code to have the correct #ifs but this also needs a fix in the analyzer to avoid the suppression.

**Comment by mavasani** [comment]
Attempting to rebase to Dev15.6 seems to have messed up the commits in this PR. Closing in favor of https://github.com/dotnet/roslyn/pull/20446
